imap/jmap_core.c

```diff
@@ -747,3 +759,248 @@ static int jmap_quota_get(jmap_req_t *req)
+                            json_integer(req->counters.notesdeletedmodseq));
+
+    if (jmap_wantprop(get->props, "mailFoldersModSeq"))
+        json_object_set_new(res, "mailFoldersModSeq",
+                            json_integer(req->counters.mailfoldersmodseq));
+    if (jmap_wantprop(get->props, "calendarFoldersModSeq"))
+        json_object_set_new(res, "calendarFoldersModSeq",
+                            json_integer(req->counters.caldavfoldersmodseq));
+    if (jmap_wantprop(get->props, "contactsFoldersModSeq"))
+        json_object_set_new(res, "contactsFoldersModSeq",
+                            json_integer(req->counters.carddavfoldersmodseq));
+    if (jmap_wantprop(get->props, "notesFoldersModSeq"))
+        json_object_set_new(res, "notesFoldersModSeq",
+                            json_integer(req->counters.notesfoldersmodseq));
+
+    if (jmap_wantprop(get->props, "mailFoldersDeletedModSeq"))
+        json_object_set_new(res, "mailFoldersDeletedModSeq",
+                            json_integer(req->counters.mailfoldersdeletedmodseq));
+    if (jmap_wantprop(get->props, "calendarFoldersDeletedModSeq"))
+        json_object_set_new(res, "calendarFoldersDeletedModSeq",
+                            json_integer(req->counters.caldavfoldersdeletedmodseq));
+    if (jmap_wantprop(get->props, "contactsFoldersDeletedModSeq"))
+        json_object_set_new(res, "contactsFoldersDeletedModSeq",
+                            json_integer(req->counters.carddavfoldersdeletedmodseq));
+    if (jmap_wantprop(get->props, "notesFoldersDeletedModSeq"))
+        json_object_set_new(res, "notesFoldersDeletedModSeq",
+                            json_integer(req->counters.notesfoldersdeletedmodseq));
+
+    if (jmap_wantprop(get->props, "quotaModSeq"))
+        json_object_set_new(res, "quotaModSeq",
+                            json_integer(req->counters.quotamodseq));
+    if (jmap_wantprop(get->props, "raclModSeq"))
+        json_object_set_new(res, "raclModSeq",
+                            json_integer(req->counters.raclmodseq));
+
+    if (jmap_wantprop(get->props, "uidValidity"))
+        json_object_set_new(res, "uidValidity",
+                            json_integer(req->counters.uidvalidity));
+
+    json_array_append_new(get->list, res);
+}
+
+static int jmap_usercounters_get(jmap_req_t *req)
+{
+    struct jmap_parser parser = JMAP_PARSER_INITIALIZER;
+    struct jmap_get get;
+    json_t *err = NULL;
+
+    /* Parse request */
+    jmap_get_parse(req, &parser, usercounters_props, /*allow_null_ids*/1,
+                   NULL, NULL, &get, &err);
+    if (err) {
+        jmap_error(req, err);
+        goto done;
+    }
+
+    /* Does the client request specific responses? */
+    if (JNOTNULL(get.ids)) {
+        json_t *jval;
+        size_t i;
+
+        json_array_foreach(get.ids, i, jval) {
+            const char *id = json_string_value(jval);
+
+            if (!strcmp(id, "singleton"))
```

Comment:
Wow, that's even more hackish than the "mail" quota id. I guess both quotas and usercounters should really go into an Account object. But let's just go with that singleton-id idiom for now.